goguen/formal-spec/utxo.tex

```diff
@@ -818,34 +775,18 @@ \subsection{Witnessing}
     {
       \var{tx}~\leteq~\fun{toGoguenTx}~{tx}_o \\~\\
       \var{txb}\leteq\txbody{tx} &
-      \var{txw}\leteq\fun{txwits}~{tx} &
-      \var{tx}~\in~\GoguenTx \\
+      \var{txw}\leteq\fun{txwits}~{tx} \\
       (utxo, \wcard, \wcard, \wcard) \leteq \var{utxoSt} \\
       \var{witsKeyHashes} \leteq \{\fun{hashKey}~\var{vk} \vert \var{vk} \in
       \dom (\txwitsVKey{txw}) \}\\~\\
-      \forall \var{validator} \in \fun{txscripts}~{txw} \cap \ScriptMSig,\\
-      \fun{runMSigScript}~\var{validator}~\var{tx}\\~\\
-      \{ h \mid (\wcard, h) \in \fun{scriptsNeeded}~\var{utxo}~\var{tx}\} ~=~ \dom (\fun{indexedScripts}~{tx}) \\
-      \forall (\var{purp}, h) \in ~\fun{scriptsNeeded}~\var{utxo}~\var{tx}, ~h\mapsto s~\in~\fun{indexedScripts}~{tx},\\
-       s \in \ScriptPlutus~\Leftrightarrow \exists r, \var{purp} \mapsto r \in \fun{findRdmr}~{tx}
+      \forall \var{validator} \in \fun{txscripts}~{txw} \cap \ScriptNative,\\
+      \fun{runNativeScript}~\var{validator}~\var{tx}\\~\\
+      \forall (\var{purp}, h) \in ~\fun{scriptsNeeded}~\var{utxo}~\var{tx}, \exists s, (h\mapsto s~\in~\fun{indexedScripts}~{tx} ~\land\\
+       s \in \ScriptNonNative~\Leftrightarrow \exists r, \var{purp} \mapsto r \in \fun{findRdmr}~{tx})
       \\~\\
```

Comment:
I am not 100% behind the idea that all outputs have an optional datum - I prefer two types of outputs I think.
But either way, we need to check in the UTXOW rule that : 

for each output `o` corresponding to an input of the transaction being processed
- `o` has a datum (or is of TxOutP type) <=> its address corresponds to a plutus script

Reply:
There isn't a type of Plutus (or non-native) scripts addresses right now though, and as far as I'm aware, we have no way of distinguishing those kinds of addresses from each other. This check would also imply that any output that has a datum but whose address does not correspond to a script would become unspendable (whereas now it just requires supplying the datum, without using it for anything). So then we would also have to ensure that a transaction doesn't generate these outputs.